logic/Rule.java

```diff
@@ -723,12 +749,23 @@ public Materialisable materialisable(ConceptMap whenConcepts, ConceptManager con
 
                     assert isa().type().label().isPresent()
                             && attribute().value().size() == 1
-                            && attribute().value().iterator().next().isValueIdentity();
+                            && attribute().value().iterator().next().predicate().equals(TypeQLToken.Predicate.Equality.EQ);
                     Label attributeTypeLabel = isa().type().label().get().properLabel();
                     AttributeType attrType = conceptMgr.getAttributeType(attributeTypeLabel.name());
                     assert attrType != null;
-                    ValueConstraint<?> value = attribute().value().iterator().next();
-                    return new Materialisable(owner, attrType, value);
+                    PredicateConstraint<?> value = attribute().value().iterator().next();
+                    if (value.isValVar()) return new Materialisable(owner, attrType, predicateConstraintForValueAssignment(whenConcepts.get(value.asValVar().value().id()).asValue()));
+                    else if (value.isConstant()) return new Materialisable(owner, attrType, value);
+                    else throw TypeDBException.of(ILLEGAL_STATE);
+                }
+
+                private PredicateConstraint.Constant<?> predicateConstraintForValueAssignment(com.vaticle.typedb.core.concept.value.Value<?> value) {
+                    if (value.isBoolean()) return new PredicateConstraint.Constant.Boolean(attribute(), TypeQLToken.Predicate.Equality.EQ, value.asBoolean().value());
+                    else if (value.isLong()) return new PredicateConstraint.Constant.Long(attribute(), TypeQLToken.Predicate.Equality.EQ, value.asLong().value());
+                    else if (value.isDouble()) return new PredicateConstraint.Constant.Double(attribute(), TypeQLToken.Predicate.Equality.EQ, value.asDouble().value());
+                    else if (value.isString()) return new PredicateConstraint.Constant.String(attribute(), TypeQLToken.Predicate.Equality.EQ, value.asString().value());
+                    else if (value.isDateTime()) return new PredicateConstraint.Constant.DateTime(attribute(), TypeQLToken.Predicate.Equality.EQ, value.asDateTime().value());
+                    else throw TypeDBException.of(ILLEGAL_STATE);
```

Comment:
Converts the ValueConcept to the Equality.EQ constraint expected by the materialiser, as it would for Constant attribute inference.